src/launch/launcher.rs

```diff
@@ -0,0 +1,112 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use super::{Measurement, Secret, Start};
+
+use crate::launch::linux::ioctl::*;
+use crate::launch::types::*;
+
+use std::io::Result;
+use std::mem::MaybeUninit;
+use std::os::unix::io::{AsRawFd, RawFd};
+
+pub struct Handle(u32);
+
+pub struct New;
+pub struct Started(Handle);
+pub struct Measured(Handle, Measurement);
+
+struct Fd<'a>(&'a dyn AsRawFd);
+
+impl AsRawFd for Fd<'_> {
+    fn as_raw_fd(&self) -> RawFd {
+        self.0.as_raw_fd()
+    }
+}
+
+pub struct Launcher<'a, T> {
+    state: T,
+    kvm: Fd<'a>,
+    sev: Fd<'a>,
+}
+
+impl<'a, T> Launcher<'a, T> {
+    pub fn pin_pages(&mut self, memory: &'a [u8]) -> Result<()> {
+        let reg_enc_region = EncryptedRegion::new(memory);
+        reg_enc_region.ioctl(&self.kvm)?;
+        Ok(())
+    }
+}
+
+impl<'a> Launcher<'a, New> {
+    pub fn new(kvm: &'a dyn AsRawFd, sev: &'a dyn AsRawFd) -> Result<Self> {
+        let mut launcher = Launcher {
+            kvm: Fd(kvm),
+            sev: Fd(sev),
+            state: New,
+        };
+
+        INIT.ioctl(&mut launcher.kvm, &mut Command::from(&launcher.sev, &Init))?;
+
+        Ok(launcher)
+    }
+
+    pub fn start(mut self, start: Start) -> Result<Launcher<'a, Started>> {
+        let mut launch_start = LaunchStart::new(&start.policy, &start.cert, &start.session);
+        LAUNCH_START.ioctl(
+            &mut self.kvm,
+            &mut Command::from_mut(&self.sev, &mut launch_start),
+        )?;
+
+        let next = Launcher {
+            state: Started(Handle(launch_start.handle)),
+            kvm: self.kvm,
+            sev: self.sev,
+        };
+
+        Ok(next)
+    }
+}
+
+impl<'a> Launcher<'a, Started> {
+    pub fn update_data(&mut self, data: &[u8]) -> Result<()> {
+        let launch_update_data = LaunchUpdateData::new(data);
+        LAUNCH_UPDATE_DATA.ioctl(
+            &mut self.kvm,
+            &mut Command::from(&self.sev, &launch_update_data),
+        )?;
+        Ok(())
+    }
+
+    pub fn measure(mut self) -> Result<Launcher<'a, Measured>> {
+        let mut measurement = unsafe { MaybeUninit::zeroed().assume_init() };
+        LAUNCH_MEASUREMENT.ioctl(
+            &mut self.kvm,
+            &mut Command::from_mut(&self.sev, &mut LaunchMeasure::new(&mut measurement)),
+        )?;
+
+        let next = Launcher {
+            state: Measured(self.state.0, measurement),
+            kvm: self.kvm,
+            sev: self.sev,
+        };
+
+        Ok(next)
+    }
+}
```

Comment:
`MaybeUninit::zeroed().assume_init()` is unsound

```suggestion
    pub fn measure(mut self) -> Result<Launcher<'a, Measured>> {
        let mut measurement = MaybeUninit::uninit();
        LAUNCH_MEASUREMENT.ioctl(
            &mut self.kvm,
            &mut Command::from_mut(&self.sev, &mut LaunchMeasure::new(&mut measurement)),
        )?;

        let next = Launcher {
            state: Measured(self.state.0, unsafe { measurement.assume_init() }),
            kvm: self.kvm,
            sev: self.sev,
        };

        Ok(next)
    }
```

Reply:
You shall only return `.assume_init()`, if the extern "C" FFI (here `libc::ioctl`) returned successfully.